### ISSUE TITLE
 Prometheus and our services throws 503 errors

### ISSUE THREAD

**Issue by Abrishges** [linkerd/linkerd2#4906]
### What is the issue?
`Linkerd check` failed and traffic flow to some of my services failed after [ Linkerd Issuer cert rotated](https://linkerd.io/2/tasks/manually-rotating-control-plane-tls-credentials/#rotating-the-identity-issuer-certificate). Prometheus pod was restarted during cert rotation 

```
NAME                                      READY   STATUS    RESTARTS   AGE
flagger-678f954485-q7j6q                  1/1     Running   0          23d
linkerd-controller-55f6f658cb-np6jn       2/2     Running   0          14d
linkerd-controller-55f6f658cb-rhxxh       2/2     Running   0          14d
linkerd-controller-55f6f658cb-vdb67       2/2     Running   0          14d
linkerd-destination-86dbc8f6c4-hxkv8      2/2     Running   0          14d
linkerd-destination-86dbc8f6c4-lcrdg      2/2     Running   0          14d
linkerd-destination-86dbc8f6c4-mljws      2/2     Running   0          14d
linkerd-destination-86dbc8f6c4-qwzdp      2/2     Running   0          14d
linkerd-destination-86dbc8f6c4-v7pkl      2/2     Running   0          14d
linkerd-grafana-7ddb5f948c-47k9x          2/2     Running   0          14d
linkerd-identity-6c58cdb5f5-l9lkx         2/2     Running   0          14d
linkerd-identity-6c58cdb5f5-s9t7c         2/2     Running   0          14d
linkerd-identity-6c58cdb5f5-vfhsj         2/2     Running   0          14d
linkerd-prometheus-bf776d569-t5br2        2/2     Running   0          8m57s
linkerd-proxy-injector-86cc796d9d-584fj   2/2     Running   0          14d
linkerd-proxy-injector-86cc796d9d-vb8fb   2/2     Running   0          14d
linkerd-proxy-injector-86cc796d9d-vzs9n   2/2     Running   0          14d
linkerd-sp-validator-6985c46b69-6z6j8     2/2     Running   0          14d
linkerd-sp-validator-6985c46b69-7ql57     2/2     Running   0          14d
linkerd-sp-validator-6985c46b69-tr9sg     2/2     Running   0          14d
linkerd-tap-5fd44d9f68-bp9cn              2/2     Running   0          14d
linkerd-tap-5fd44d9f68-hllz6              2/2     Running   0          14d
linkerd-tap-5fd44d9f68-m29bw              2/2     Running   0          14d
linkerd-web-6bb6886855-tjmtj              2/2     Running   0          14d
```

### How can it be reproduced?
As of now, I've been unable to reproduce it. I ran issuer cert in other clusters, the behavior doesn't present.

### Event after Issuer rotation 
```kubectl get events --field-selector reason=IssuerUpdated -n linkerd
LAST SEEN   TYPE     REASON          OBJECT                        MESSAGE
102s        Normal   IssuerUpdated   deployment/linkerd-identity   Updated identity issuer
98s         Normal   IssuerUpdated   deployment/linkerd-identity   Updated identity issuer
62s         Normal   IssuerUpdated   deployment/linkerd-identity   Updated identity issuer  
```
### Prometheus log 

````
time="2020-08-20T23:49:45Z" level=info msg="running version stable-2.7.1"
[     0.11638780s]  INFO linkerd2_proxy: Admin interface on 0.0.0.0:4191
[     0.11777381s]  INFO linkerd2_proxy: Inbound interface on 0.0.0.0:4143
[     0.11789281s]  INFO linkerd2_proxy: Outbound interface on 127.0.0.1:4140
[     0.11795081s]  INFO linkerd2_proxy: Tap interface on 0.0.0.0:4190
[     0.11800281s]  INFO linkerd2_proxy: Local identity is linkerd-prometheus.linkerd.serviceaccount.identity.linkerd.cluster.local
[     0.11807481s]  INFO linkerd2_proxy: Identity verified via linkerd-identity.linkerd.svc.cluster.local:8080 (linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.11813781s]  INFO linkerd2_proxy: Destinations resolved via linkerd-dst.linkerd.svc.cluster.local:8086 (linkerd-destination.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.12391285s]  INFO inbound: linkerd2_app_inbound: Serving listen.addr=0.0.0.0:4143
[     0.110020054s]  INFO daemon:identity: linkerd2_app: Certified identity: linkerd-prometheus.linkerd.serviceaccount.identity.linkerd.cluster.local
[    35.254456607s]  WARN outbound:accept{peer.addr=172.21.5.126:40210}:source{target.addr=172.21.0.8:9090}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[    36.14091009s]  WARN outbound:accept{peer.addr=172.21.5.126:37162}:source{target.addr=172.21.0.8:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[    50.254551630s]  WARN outbound:accept{peer.addr=172.21.5.126:40210}:source{target.addr=172.21.0.8:9090}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[    51.13924731s]  WARN outbound:accept{peer.addr=172.21.5.126:37162}:source{target.addr=172.21.0.8:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[   974.697162753s]  WARN outbound:accept{peer.addr=172.21.5.126:46360}:source{target.addr=172.21.21.51:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: request timed out
[  1012.334721788s]  WARN outbound:accept{peer.addr=172.21.5.126:59142}:source{target.addr=172.21.18.37:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[  1012.416644670s]  WARN outbound:accept{peer.addr=172.21.5.126:55138}:source{target.addr=172.21.21.37:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[  1019.536720435s]  WARN outbound:accept{peer.addr=172.21.5.126:42302}:source{target.addr=172.21.17.107:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: request timed out
[  1069.893730212s]  WARN outbound:accept{peer.addr=172.21.5.126:44674}:source{target.addr=172.21.15.132:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: request timed out
[  1070.640813202s]  WARN outbound:accept{peer.addr=172.21.5.126:43914}:source{target.addr=172.21.17.103:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: request timed out
[  1073.916726399s]  WARN outbound:accept{peer.addr=172.21.5.126:43614}:source{target.addr=172.21.15.253:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: request timed out
[ 64770.844834946s]  WARN outbound:accept{peer.addr=172.21.5.126:51206}:source{target.addr=172.21.19.51:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[ 64833.562333380s]  WARN outbound:accept{peer.addr=172.21.5.126:33420}:source{target.addr=172.21.21.161:4191}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
[ 77702.780856541s]  WARN outbound:accept{peer.addr=172.21.5.126:37236}:source{target.addr=172.17.59.39:80}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: request timed out
[ 78363.564560027s]  WARN outbound:accept{peer.addr=172.21.5.126:39982}:source{target.addr=172.17.59.39:80}: linkerd2_app_core::errors: Failed to proxy request: connection error: Connection timed out (os error 110)
````

#### `linkerd check` output

```kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust roots are using supported crypto algorithm
√ trust roots are within their validity period
√ trust roots are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust root

linkerd-identity-data-plane
---------------------------
√ data plane proxies certificate match CA

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
× [prometheus] control plane can talk to Prometheus
    Error calling Prometheus from the control plane: server_error: server error: 503
    see https://linkerd.io/checks/#l5d-api-control-api for hints

```
### logs from controller pods 
```
kubectl logs -c linkerd-proxy linkerd-controller-74448d4d64-nlccx -n linkerd
time="2020-08-10T21:31:06Z" level=info msg="running version stable-2.7.1"
[     0.11111911s]  INFO linkerd2_proxy: Admin interface on 0.0.0.0:4191
[     0.11145611s]  INFO linkerd2_proxy: Inbound interface on 0.0.0.0:4143
[     0.11152411s]  INFO linkerd2_proxy: Outbound interface on 127.0.0.1:4140
[     0.11157311s]  INFO linkerd2_proxy: Tap interface on 0.0.0.0:4190
[     0.11162411s]  INFO linkerd2_proxy: Local identity is linkerd-controller.linkerd.serviceaccount.identity.linkerd.cluster.local
[     0.11168511s]  INFO linkerd2_proxy: Identity verified via linkerd-identity.linkerd.svc.cluster.local:8080 (linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.11174411s]  INFO linkerd2_proxy: Destinations resolved via linkerd-dst.linkerd.svc.cluster.local:8086 (linkerd-destination.linkerd.serviceaccount.identity.linkerd.cluster.local)
[     0.11692607s]  INFO inbound: linkerd2_app_inbound: Serving listen.addr=0.0.0.0:4143
[     0.48824911s]  INFO daemon:identity: linkerd2_app: Certified identity: linkerd-controller.linkerd.serviceaccount.identity.linkerd.cluster.local
[ 60493.899598588s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
[343593.565154583s]  WARN dst: linkerd2_reconnect::service: Service failed error=channel closed
[934041.797573538s]  WARN outbound:accept{peer.addr=172.21.15.9:46580}:source{target.addr=172.17.77.35:9090}: linkerd2_app_core::errors: Failed to proxy request: request timed out
[934278.634996170s]  WARN outbound:accept{peer.addr=172.21.15.9:46580}:source{target.addr=172.17.77.35:9090}: linkerd2_app_core::errors: Failed to proxy request: request timed out
[934642.349165418s]  WARN outbound:accept{peer.addr=172.21.15.9:46580}:source{target.addr=172.17.77.35:9090}: linkerd2_app_core::errors: Failed to proxy request: request timed out
```
Prometheus pods restart didn’t fix the issue. Port forward for linkerd-prometheus page works. After reading [similar github issue](https://github.com/linkerd/linkerd2/issues/3819#issuecomment-571612111), killed controller pods which fixed Prometheus issue. My services continue throw 503. 

### logs from Identity pods 
```
8/21/2020 3:27:52 PM  linkerd-identity-8547445c87-xx769      {2020-08-21T22:24:52.866068532Z [    48.40397513s]  WARN inbound:accept{peer.addr=172.21.15.162:55382}:source{target.addr=172.21.14.179:8080}: linkerd2_app_core::errors: Failed ... 8/21/2020 3:27:23 PM  linkerd-identity-8547445c87-dxfph      {2020-08-21T22:25:53.967176323Z [    69.231548883s]  WARN inbound:accept{peer.addr=172.21.15.162:43216}:source{target.addr=172.21.15.31:8080}: linkerd2_app_core::errors: Failed ... 8/21/2020 3:26:54 PM  linkerd-prometheus-6fcfb75b56-bhfpw    {2020-08-21T17:07:00.527772461Z [    15.220331303s]  WARN outbound:accept{peer.addr=172.21.5.151:40724}:source{target.addr=172.21.2.23:9090}: linkerd2_app_core::errors: Failed t... 8/21/2020 3:26:54 PM  linkerd-identity-8547445c87-9sn8x      {2020-08-21T22:25:23.062675766Z [    47.594859589s]  WARN inbound:accept{peer.addr=172.21.15.162:33126}:source{target.addr=172.21.15.92:8080}: linkerd2_app_core::errors: Failed ... 8/21/2020 12:40:51 PM linkerd-web-5449f69d8c-mdlr6           {2020-08-11T14:19:08.065868406Z [ 60493.890611905s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed, 2020-08-21T19:40:51.992221664Z [9437... 8/11/2020 7:19:47 AM  linkerd-sp-validator-ccf694689-rpplw   2020-08-11T14:19:47.288660053Z [ 60493.366967770s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:43 AM  linkerd-proxy-injector-fc4f958d6-568gj 2020-08-11T14:19:43.075044261Z [ 60493.886023905s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:34 AM  linkerd-sp-validator-ccf694689-n9c7b   2020-08-11T14:19:34.009291899Z [ 60494.25051798s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:20 AM  linkerd-proxy-injector-fc4f958d6-q9xk8 2020-08-11T14:19:20.091811554Z [ 60493.821667393s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:20 AM  linkerd-tap-77bf555b7b-g9949           2020-08-11T14:19:20.026615511Z [ 60493.994400958s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:19 AM  linkerd-sp-validator-ccf694689-rjwtd   2020-08-11T14:19:19.29775263Z [ 60493.366676273s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:06 AM  linkerd-grafana-6c87b754c9-wzpq7       {2020-08-10T21:30:54.68005741Z [     1.850826029s]  WARN inbound:accept{peer.addr=172.21.15.59:41538}:source{target.addr=172.21.15.95:3000}: linkerd2_app_core::errors: Failed to...
```
### How I be able to recover our services ? 
Our services recovered after Killing Identity and destination pods. 

### Environment

- Kubernetes Version:  v1.16.10
- Cluster Environment:  AKS
- Host OS: Ubuntu 16.04
- Linkerd version: 2.7.1

### Additional context

- We have problems when a nodes outage and waiting for fix
- AKS removed the `config.linkerd.io/admission-webhooks` label [from the kube-system ](https://github.com/linkerd/linkerd2/issues/4754) and waiting for future stable release to change `failurePolicy `to `ignore `

 - Linkerd check before running Issuer cert 
```
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-identity
----------------
√ certificate config is valid
√ trust roots are using supported crypto algorithm
√ trust roots are within their validity period
√ trust roots are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust root

linkerd-api
-----------
√ control plane pods are ready
√ control plane self-check
√ [kubernetes] control plane can talk to Kubernetes
√ [prometheus] control plane can talk to Prometheus
√ tap api service is running

linkerd-version
---------------
√ can determine the latest version
‼ cli is up-to-date
    is running version 2.7.1 but the latest stable version is 2.8.1
    see https://linkerd.io/checks/#l5d-version-cli for hints

control-plane-version
---------------------
‼ control plane is up-to-date
    is running version 2.7.1 but the latest stable version is 2.8.1
    see https://linkerd.io/checks/#l5d-version-control for hints
√ control plane and cli versions match

linkerd-ha-checks
-----------------
‼ pod injection disabled on kube-system
    kube-system namespace needs to have the label config.linkerd.io/admission-webhooks: disabled if HA mode is enabled
    see https://linkerd.io/checks/#l5d-injection-disabled for hints
```

**Comment by ihcsim** [comment]
At which step did the 503s start appearing? Right after `linkerd upgrade  --identity-issuer-certificate-file=./issuer-new.crt --identity-issuer-key-file=./issuer-new.key | kubectl apply -f -`?

Looking at your Prometheus proxy logs, seems like it's getting a bunch of outbound `Connection refused` from other pods it's trying to scrape. Are those Linkerd control plane pods?

> Our services recovered after Killing Identity and destination pods.

Do you have any logs from the `identity` service?

**Comment by ihcsim** [comment]
I can't reproduce the errors you are seeing on GKE, following the steps in the rotating cert documentation. One thing I did notice is that the step on rotating the *issuer cert* should not restart your Prometheus instance, because the only thing that is changed is the identity secret.

If you have more info that you think will be helpful with repro, feel free to add them here. Thanks.

**Comment by Abrishges** [comment]
> At which step did the 503s start appearing? Right after `linkerd upgrade --identity-issuer-certificate-file=./issuer-new.crt --identity-issuer-key-file=./issuer-new.key | kubectl apply -f -`?
> 
> Looking at your Prometheus proxy logs, seems like it's getting a bunch of outbound `Connection refused` from other pods it's trying to scrape. Are those Linkerd control plane pods?
> 
> > Our services recovered after Killing Identity and destination pods.
> 
> Do you have any logs from the `identity` service?

**Comment by Abrishges** [comment]
> At which step did the 503s start appearing? Right after `linkerd upgrade --identity-issuer-certificate-file=./issuer-new.crt --identity-issuer-key-file=./issuer-new.key | kubectl apply -f -`?
> 
> Looking at your Prometheus proxy logs, seems like it's getting a bunch of outbound `Connection refused` from other pods it's trying to scrape. Are those Linkerd control plane pods?
> 
> > Our services recovered after Killing Identity and destination pods.
> 
> Do you have any logs from the `identity` service?

Prometheus pods restarted due to small configuration change when `linkerd upgrade` command is applied. 
Prometheus proxy logs get outbound `connection refused` errors from linkerd control pods 
### error/warn logs from linkerd namespace including identity
```
8/21/2020 3:27:52 PM  linkerd-identity-8547445c87-xx769      {2020-08-21T22:24:52.866068532Z [    48.40397513s]  WARN inbound:accept{peer.addr=172.21.15.162:55382}:source{target.addr=172.21.14.179:8080}: linkerd2_app_core::errors: Failed ... 
8/21/2020 3:27:23 PM  linkerd-identity-8547445c87-dxfph      {2020-08-21T22:25:53.967176323Z [    69.231548883s]  WARN inbound:accept{peer.addr=172.21.15.162:43216}:source{target.addr=172.21.15.31:8080}: linkerd2_app_core::errors: Failed ... 
8/21/2020 3:26:54 PM  linkerd-prometheus-6fcfb75b56-bhfpw    {2020-08-21T17:07:00.527772461Z [    15.220331303s]  WARN outbound:accept{peer.addr=172.21.5.151:40724}:source{target.addr=172.21.2.23:9090}: linkerd2_app_core::errors: Failed t...
8/21/2020 3:26:54 PM  linkerd-identity-8547445c87-9sn8x      {2020-08-21T22:25:23.062675766Z [    47.594859589s]  WARN inbound:accept{peer.addr=172.21.15.162:33126}:source{target.addr=172.21.15.92:8080}: linkerd2_app_core::errors: Failed ...
8/21/2020 12:40:51PM linkerd-web-5449f69d8c-mdlr6           {2020-08-11T14:19:08.065868406Z [ 60493.890611905s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed, 2020-08-21T19:40:51.992221664Z [9437... 
8/11/2020 7:19:47 AM  linkerd-sp-validator-ccf694689-rpplw   2020-08-11T14:19:47.288660053Z [ 60493.366967770s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:43 AM  linkerd-proxy-injector-fc4f958d6-568gj 2020-08-11T14:19:43.075044261Z [ 60493.886023905s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:34 AM  linkerd-sp-validator-ccf694689-n9c7b   2020-08-11T14:19:34.009291899Z [ 60494.25051798s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:20 AM  linkerd-proxy-injector-fc4f958d6-q9xk8 2020-08-11T14:19:20.091811554Z [ 60493.821667393s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:20 AM  linkerd-tap-77bf555b7b-g9949           2020-08-11T14:19:20.026615511Z [ 60493.994400958s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:19 AM  linkerd-sp-validator-ccf694689-rjwtd   2020-08-11T14:19:19.29775263Z [ 60493.366676273s]  WARN daemon:identity: linkerd2_reconnect::service: Service failed error=channel closed
8/11/2020 7:19:06 AM  linkerd-grafana-6c87b754c9-wzpq7       {2020-08-10T21:30:54.68005741Z [     1.850826029s]  WARN inbound:accept{peer.addr=172.21.15.59:41538}:source{target.addr=172.21.15.95:3000}: linkerd2_app_core::errors: Failed to...
```

**Comment by ihcsim** [comment]
Is this reproducible on your cluster?
